### DISCUSSION TITLE
`table` is special cased to `{any:any}`, but should it allow any record?

### DISCUSSION THREAD

**Discussion by mtdowling** [teal-language/tl#774] [General]
`table` is an alias for `{any:any}` based on this test case: https://github.com/teal-language/tl/blob/master/spec/subtyping/table_spec.lua#L3. I assumed that `record` is a subtype of `table`, allowing me to pass in any record, `{any}`, and `{any:any}`. If it's relevant, my use case is that I want to accept any value that could have a metatable.

```
local function f(x: table) -- < wanted this to accept records too
end

local record R
end

local r: R = {}

f(r)
```

Fails with:

```
argument 1: can't match a record to a map with non-string keys
```

If it's desirable to distinguish between records and tables, then if there was a type like `record` that accepts any record, then I suppose I could do `table | record`.

**Comment by hishamhm** [comment]
The `table` notation was a very early addition back from when Teal was still bootstrapping itself. It ended up kind of sticking around. I don't even use myself anymore, and just use `{any:any}` in similar situations.

Mixing such general table types with records is often a sign of something tricky going on. When such use of the table is read-only, of course things are safe, but often we're using `{any:any}` to do some smart manipulation of the data (e.g. copying any object generically, etc.) — it's easy to end up breaking some of the compiler's assumptions when doing that (e.g. setting a record field with a value of an invalid type). So I didn't want to make it too "natural" for such dynamic uses to pop up in the language. I think it's a good thing that whenever something unusual (in statically-typed terms) like that happens, the code has a cast somewhere that acts as a warning to the reader.

So, right now the alternative to do such general-table manipulation on record is really to do `f(r as {any:any})`.

In my experience, this has been a good thing: when I'm hunting my own bugs in Teal, the presence of `as` is often a good hint that I might have tripped something up with my tables.
